[JSC] Iterator helpers incorrectly close iterators on early errors

Constellation opened the pull request:
#### d6f3d9c7c304a8d5768e4a5ff4f1ce27e2dbfa47
<pre>
[JSC] Iterator helpers incorrectly close iterators on early errors
<a href="https://bugs.webkit.org/show_bug.cgi?id=291195">https://bugs.webkit.org/show_bug.cgi?id=291195</a>
<a href="https://rdar.apple.com/148774612">rdar://148774612</a>

Reviewed by Keith Miller and Sosuke Suzuki.

NaN / Integer checks should be done without iterator closing.
We also fix iterator helpers&apos; closing timing for the other methods as
well.

* JSTests/stress/iterator-helper-close-timing.js: Added.
(shouldBe):
(shouldThrow):
* Source/JavaScriptCore/builtins/JSIteratorPrototype.js:

Canonical link: <a href="https://commits.webkit.org/293842@main">https://commits.webkit.org/293842@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc43c36667ed91a3bf4e88e4f0521fbc77133682

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105114 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50567 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28136 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76120 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33202 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102993 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15224 "Found 1 new test failure: media/video-aspect-ratio.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90299 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56479 "Found 2 new API test failures: /TestWebKit:WebKit.UserMediaBasic, /WPE/TestWebKitWebView:/webkit/WebKitWebView/is-playing-audio (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15029 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-contain/contain-size-021.html imported/w3c/web-platform-tests/css/css-position/sticky/position-sticky-fixed-ancestor-iframe.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8289 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49936 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92644 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84954 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8374 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107474 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98593 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19869 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85073 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27462 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86499 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84597 "Found 6 new API test failures: /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, /TestWebKit:WebKit.OnDeviceChangeCrash, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/mouse-target, /TestWebKit:WebKit.UserMediaBasic, /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/custom-container-destroyed, /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/default (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29275 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6999 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20935 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16285 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27036 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32264 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122219 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26847 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34122 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30163 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28406 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki reviewed (commented):
This changes introduces new fails on test262

(Here is my local test result)
```
! NEW FAIL test/built-ins/Iterator/prototype/drop/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/drop/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/find/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/find/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/some/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/some/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/take/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/take/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/filter/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/filter/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/reduce/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/reduce/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/map/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/map/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/every/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/every/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/flatMap/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/flatMap/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/forEach/argument-validation-failure-closes-underlying.js (strict mode)
Exception: Test262Error: Expected SameValue(«false», «true») to be true

! NEW FAIL test/built-ins/Iterator/prototype/forEach/argument-validation-failure-closes-underlying.js (default)
Exception: Test262Error: Expected SameValue(«false», «true») to be true
```

kmiller68 reviewed (approved):
r=me, we need macros for our builtins... lol

Constellation committed:
[JSC] Iterator helpers incorrectly close iterators on early errors
https://bugs.webkit.org/show_bug.cgi?id=291195
rdar://148774612

Reviewed by Keith Miller and Sosuke Suzuki.

NaN / Integer checks should be done without iterator closing.
We also fix iterator helpers' closing timing for the other methods as
well.

* JSTests/stress/iterator-helper-close-timing.js: Added.
(shouldBe):
(shouldThrow):
* Source/JavaScriptCore/builtins/JSIteratorPrototype.js:

Canonical link: https://commits.webkit.org/293842@main
